python/tank_vendor/shotgun_authentication/user.py

```diff
@@ -0,0 +1,392 @@
+        """
+        credentials = session_cache.get_session_data(host)
+        if credentials:
+            return SessionUser(
+                host=host,
+                http_proxy=http_proxy,
+                **credentials
+            )
+        else:
+            return None
+
+    @staticmethod
+    def from_dict(representation):
+        """
+        Creates a user from a dictionary.
+
+        :param payload: Dictionary with the user information.
+
+        :returns: A SessionUser instance.
+        """
+
+        return SessionUser(**representation)
+
+    def to_dict(self):
+        """
+        Converts the user into a dictionary object.
+
+        :returns: A dictionary with all the attributes of the user.
+        """
+        data = super(SessionUser, self).to_dict()
+        data["login"] = self._login
+        data["session_token"] = self._session_token
+        data["is_volatile"] = self._is_volatile
+        return data
+
+
+class ScriptUser(ShotgunUser):
+    """
+    User that authenticates to the Shotgun server using a api name and api key.
+    """
+    def __init__(self, host, api_script, api_key, http_proxy):
+        """
+        Constructor.
+
+        :param host: Host for this Shotgun user.
+        :param api_script: API script name.
+        :param api_key: API script key.
+        :param http_proxy: HTTP proxy to use with this host. Defaults to None.
+        """
+        super(ScriptUser, self).__init__(host, http_proxy)
+
+        self._api_script = api_script
+        self._api_key = api_key
+
+    def create_sg_connection(self):
+        """
+        Creates a Shotgun instance using the script user's credentials.
+
+        :returns: A Shotgun instance.
+        """
+        return _shotgun_instance_factory(
+            self._host,
+            script_name=self._api_script,
+            api_key=self._api_key,
+            http_proxy=self._http_proxy,
+            user=self
+        )
+
+    def get_script(self):
+        """
+        Returns the script user name.
+
+        :returns: The script user name.
+        """
+        return self._api_script
+
+    def get_key(self):
+        """
+        Returns the script user key.
+
+        :returns: The script user key.
+        """
+        return self._api_key
+
+    def to_dict(self):
+        """
+        Converts the user into a dictionary object.
+
+        :returns: A dictionary with all the attributes of the user.
+        """
+        data = super(ScriptUser, self).to_dict()
+        data["api_script"] = self._api_script
+        data["api_key"] = self._api_key
+        return data
+
+    @staticmethod
+    def from_dict(representation):
+        """
+        Creates a user from a dictionary.
+
+        :param payload: Dictionary with the user information.
+
+        :returns: A ScriptUser instance.
+        """
+        return ScriptUser(**representation)
+
+
+def is_script_user(user):
```

Comment:
I would love if we could have a single interface which wrapped both these object, and we didn't have to do this RTTI. But maybe it's not practical!